### PR TITLE
Certificate events

### PR DESCRIPTION
**Pull request by mattdean-digicatapult**
Adds events to certificates which are tracked against the block time, for example:

```
{
  "hydrogen_owner": "Heidi",
  "energy_owner": "Emma",
  "regulator": "Reginald",
  "hydrogen_quantity_wh": 3941000,
  "original_token_id": 1,
  "latest_token_id": 4,
  "commitment": "cdc4c24cab71f08f5884b454193cc6e8",
  "commitment_salt": "359f44139834600b403728848d7e24e3",
  "production_start_time": "2023-07-14T00:55:24.645Z",
  "production_end_time": "2023-07-21T00:55:24.645Z",
  "energy_consumed_wh": 5911500,
  "id": "cd30be19-b5ca-48ec-97b7-4265a2f41e19",
  "state": "issued",
  "created_at": "2024-03-21T09:53:42.198Z",
  "updated_at": "2024-03-21T09:54:19.869Z",
  "embodied_co2": 903918,
  "revocation_reason": null,
  "events": [
    {
      "event": "initiated",
      "occurred_at": "2024-03-21T09:53:48.001Z"
    },
    {
      "event": "issued",
      "occurred_at": "2024-03-21T09:54:06.001Z"
    }
  ]
}
```

Note I've left some comments below on some changes that might look a bit foreign to this PR

### CODE REVIEW

File: src/controllers/v1/certificate/index.ts [hunk 1]
```diff
@@ -294,8 +315,7 @@ export class CertificateController extends Controller {
       transaction_type: 'initiate_cert',
     })
     if (!transaction) throw new InternalServerError('Transaction must exist')
-
-    this.node.submitRunProcess(extrinsic, (state: TransactionState) =>
+    await this.node.submitRunProcess(extrinsic, (state: TransactionState) =>
```
Comment (mattdean-digicatapult) [inline src/controllers/v1/certificate/index.ts:318]: These `await`s have been put in to protext against a potential race condition for the init script. They only wait for the transaction submission and not it's inclusion or finalisation so won't affect anything but they do guarentee the nonce from the tx is in the transaction pool

File: src/lib/indexer/index.ts [hunk 2]
```diff
@@ -220,7 +220,7 @@ export default class Indexer {
         for (const [, certificate] of changeSet.certificates) {
           switch (certificate.type) {
             case 'insert': {
-              const { type, id, ...record } = certificate
+              const { type, ...record } = certificate
```
Comment (mattdean-digicatapult) [inline src/lib/indexer/index.ts:223]: this was a bug which prevented us linking a `certificate_event` to a `certificate`

File: test/helpers/chainTest.ts [hunk 3]
```diff
@@ -34,17 +35,17 @@ const submitRunProcessExtrinsicAndSeal = async (
 }
 
 export const withAppAndIndexer = (context: { app: Express; indexer: Indexer }) => {
```
Comment (mattdean-digicatapult) [inline test/helpers/chainTest.ts:37]: basically none of this worked. We somehow lost the trim of the blockhash so the insert just failed. Also this wasn't running for each test so one failure was causing all the tests to fail

**Comment by rmlearney-digicatapult** [comment]
Are `created_at` and `updated_at` relevant anymore?

File: src/lib/indexer/__tests__/eventProcessor.test.ts [hunk 4]
```diff
@@ -129,7 +165,17 @@ describe('eventProcessor', function () {
...
+      )
+      const events = [...(result.certificateEvents || [])]
+      expect(events.length).to.equal(1)
+      const event = events[0]
+      expect(event[1]).to.deep.equal({
+        type: 'insert',
+        id: event[0],
+        certificate_id: 'caa699b7-b0b6-4e0e-ac15-698b7b1f6541',
+        occurred_at: blockTime,
+        event: 'issued',
```
Comment (dc-andysign) [inline src/lib/indexer/__tests__/eventProcessor.test.ts:178]: Does this mean in our React frontend , in the component that renders the timestamp we will have to change from `updated_at` to this ?

**Review by dc-andysign — APPROVED** [review]
👍

**Comment by mattdean-digicatapult** [comment]
> Are `created_at` and `updated_at` relevant anymore?

Yes these represent when changes were made loacally. This is especially relevant if there is a delay between block generation and the event being processed. Also `created_at` and `updated_at` columns are typically good practice to have from a tracing PoV

File: src/lib/indexer/__tests__/eventProcessor.test.ts (hunk 4 above)
Comment (mattdean-digicatapult) [reply src/lib/indexer/__tests__/eventProcessor.test.ts:178]: at some point yes

**Pull request merged**